cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,567 @@
+allowing the HPA also to faster provide next scaling decisions.
+
+Example buffer:
+
+```
+apiversion: autoscaling.x-k8s.io/v1alpha1
+kind: Buffer
+metadata:
+  name: my-deployment-buffer
+  namespace: my-namespace
+spec:
+  targetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: my-deployment
+  type: buffer.x-k8s.io/active-capacity
+  recreationStrategy: keep-empty
+  replicaBasedBufferSpec:
+    replicasCountPercentage: 10
+    minCount: 1
+```
+
+### Story 2
+
+The user is a part of an admin team owning the CI/CD pipeline. The cluster is
+used for running one off test tasks that should be completed promptly. However
+the cluster is scaling up and down making the tasks wait for adding a new node.
+The admin team wants to keep spare space in the cluster so that the tasks
+schedule quickly.
+
+Example buffer:
+
+```
+apiversion: autoscaling.x-k8s.io/v1alpha1
+kind: Buffer
+metadata:
+  name: testing-capacity
+  namespace: my-namespace
+spec:
+  targetRef:
+    apiVersion: apps/v1
+    kind: CustomComputeClass
+    name: my-custom-class
+  type: buffer.x-k8s.io/active-capacity
+  recreationStrategy: keep-empty
+  nodeClassBufferSpec:
+    totalCpu: 40
+    totalMem: 5120Mi
+```
+
+Note: in the example the total size would be split into smaller chunks with some
+default shape. If a non-standard shape is needed it can be further defined in
+the `spec.nodeClassBufferSpec.perChunkResources`
+
+## Notes/Constraints/Caveats (Optional)
+
+*   The Buffer implementation assumes that the pods run as a part of scale
+    subresource are homogeneous.
+*   The initial implementation would not have an option to specify
+    ResourceClaims: the DRA integration is still in alpha phase and there are
+    implementation tasks left in CA to make such integration possible. This is
+    planned as a part of a next milestone.
+
+## Risks and Mitigations
+
+*   The implementation of buffers will introduce additional “virtual” pods to be
+    handled by the autoscalers. In case the buffer chunks were very small the
+    number of processed pods may grow - we do not expect that buffers would
+    introduce more chunks than the number of pods that user may want to run.
+    Since these are only virtual they will introduce less overhead in the
+    cluster than balloon pods/deployments (no need to preempt etc).
+
+# Design Details
+
+## Proposed API
+
+```go
+type Buffer struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+    Spec   BufferSpec
+    Status BufferStatus
+}
+
+type BufferSpec struct {
+    // Optional pointer to an object that defines how to interpret this buffer.
+    // Depending on the type of referenced object one of the configurations below will be used:
+    //   - for a crd implementing scale subresource (like deployment, job, replicaset)
+    //     the ReplicaBasedBufferSpec will be used
+    //   - for GKE CCC or Karpenter NodePool the NodeClassBufferSpec will be used
```

Comment:
For node-centric headroom, I suspect that this will be implemented in the NodePool API in Karpenter, and not part of this buffer API.